src/eckit/runtime/Main.cc

```diff
@@ -24,21 +25,27 @@
 #include "eckit/system/LibraryManager.h"
 #include "eckit/system/SystemInfo.h"
 #include "eckit/thread/AutoLock.h"
-#include "eckit/thread/Mutex.h"
 #include "eckit/thread/StaticMutex.h"
 #include "eckit/utils/Translator.h"
 
 namespace eckit {
 
 //----------------------------------------------------------------------------------------------------------------------
 
+namespace {
+// atomic provides implicit lock for finalised_ only
+std::atomic_bool finalised_ {false};
+}  // namespace
+
 static StaticMutex local_mutex;
 static Main* instance_ = nullptr;
 
 //----------------------------------------------------------------------------------------------------------------------
 
 Main::Main(int argc, char** argv, const char* homeenv) :
     taskID_(-1), argc_(argc), argv_(argv), home_("/"), debug_(false) {
+ 
+    std::atexit(finalise);
```

Comment:
Be aware that this may still interleave with static object destruction, see https://en.cppreference.com/w/cpp/utility/program/exit